### PR TITLE
Configure pause container

### PR DESCRIPTION
**Pull request by samuelkarp**
### Summary
config: Make pause image name + tag configurable

### Implementation details
Pause image name and tag are made configurable only via the config file as it's not a generally useful feature and we do not expect it to be customized under normal circumstances.

### Testing
- [x] Builds on Linux (`make release`)
- [x] Builds on Windows (`go build -out amazon-ecs-agent.exe ./agent`)
- [x] Unit tests on Linux (`make test`) pass
- [x] Unit tests on Windows (`go test -timeout=25s ./agent/...`) pass
- [x] Integration tests on Linux (`make run-integ-tests`) pass
- [x] Integration tests on Windows (`.\scripts\run-integ-tests.ps1`) pass
- [x] Functional tests on Linux (`make run-functional-tests`) pass
- [x] Functional tests on Windows (`.\scripts\run-functional-tests.ps1`) pass

New tests cover the changes: yes

### Description for the changelog
None

### Licensing
This contribution is under the terms of the Apache 2.0 License: yes, Amazon employee

### CODE REVIEW

**Comment by samuelkarp** [comment]
I'm seeing a couple test failures that look new to me, but I think are unrelated to my commit (they may already be present on the `enis` branch):

* Data race in `TestDoStartHappyPath`/`TestDoStartTaskENIHappyPath` around the package-global variable `singleUpdater` in `agent/acs/update_handler/updater.go`.
* Timeout in `TestManyDones` (inconsistent; see this both passing and timing out).
* Data race in `TestVolumesFromRO`; not sure why this is failing yet.

**Comment by samuelkarp** [comment]
@richardpen let me know that one of the races is already known: https://github.com/aws/amazon-ecs-agent/issues/922

File: agent/api/task.go [hunk 1]
```diff
@@ -243,7 +243,7 @@ func (task *Task) isNetworkModeVPC() bool {
 	return true
 }
 
-func (task *Task) addNetworkResourceProvisioningDependency() {
+func (task *Task) addNetworkResourceProvisioningDependency(cfg *config.Config) {
 	// TODO check networking mode for the task before doing this
```
Comment (samuelkarp) [inline agent/api/task.go]: Done

File: agent/config/config.go [hunk 2]
```diff
@@ -196,33 +196,33 @@ func (cfg *Config) trimWhitespace() {
 }
 
 func fileConfig() (Config, error) {
-	config_file := utils.DefaultIfBlank(os.Getenv("ECS_AGENT_CONFIG_FILE_PATH"), "/etc/ecs_container_agent/config.json")
-	config := Config{}
+	fileName := utils.DefaultIfBlank(os.Getenv("ECS_AGENT_CONFIG_FILE_PATH"), "/etc/ecs_container_agent/config.json")
```
Comment (aaithal) [inline agent/config/config.go]: While you are here, can you const `"/etc/ecs_container_agent/config.json"` as well?

File: agent/config/config.go [hunk 3]
```diff
@@ -485,28 +485,28 @@ func (config *Config) validateAndOverrideBounds() error {
...
+	cfg.platformOverrides()
 
 	return nil
 }
 
 // String returns a lossy string representation of the config suitable for human readable display.
 // Consequently, it *should not* return any sensitive information.
-func (config *Config) String() string {
-	return fmt.Sprintf("Cluster: %v, Region: %v, DataDir: %v, Checkpoint: %v, AuthType: %v, UpdatesEnabled: %v, DisableMetrics: %v, ReservedMem: %v, TaskCleanupWaitDuration: %v, DockerStopTimeout: %v", config.Cluster, config.AWSRegion, config.DataDir, config.Checkpoint, config.EngineAuthType, config.UpdatesEnabled, config.DisableMetrics, config.ReservedMemory, config.TaskCleanupWaitDuration, config.DockerStopTimeout)
+func (cfg *Config) String() string {
```
Comment (aaithal) [inline agent/config/config.go]: We can choose to do it in a separate PR, but we should get rid of this and instead replicate the functionality in `_unix.go` and `_windows.go` files so that new config fields for linux are also printed when this is invoked.

File: agent/config/config_unix.go [hunk 4]
```diff
@@ -41,8 +41,14 @@ func DefaultConfig() Config {
...
-		PauseContainerTag:           PauseContainerTag,
+		PauseContainerImageName:     DefaultPauseContainerImageName,
+		PauseContainerTag:           DefaultPauseContainerTag,
 	}
 }
 
-func (config *Config) platformOverrides() {}
+func (cfg *Config) platformOverrides() {}
+
+func (cfg *Config) ShouldLoadPauseContainerTarball() bool {
```
Comment (aaithal) [inline agent/config/config_unix.go]: lint: missing comments

File: agent/config/types.go [hunk 5]
```diff
@@ -154,7 +154,14 @@ type Config struct {
 	// PauseContainerTarballPath is the path to the pause container tarball
 	PauseContainerTarballPath string
 
-	// PauseContainer is the tag for the pause container image
+	// PauseContainerImageName is the name for the pause container image.
+	// Setting this value to be different from the default will disable loading
+	// the image from the tarball; the referenced image must already be loaded.
+	PauseContainerImageName string
+
+	// PauseContainer is the tag for the pause container image.
```
Comment (aaithal) [inline agent/config/types.go]: nit: `PauseContainerTag`

File: agent/config/config_unix.go (hunk 4 above)
Comment (samuelkarp) [reply agent/config/config_unix.go]: Done

File: agent/config/types.go (hunk 5 above)
Comment (samuelkarp) [reply agent/config/types.go]: Done

File: agent/config/config.go (hunk 2 above)
Comment (samuelkarp) [reply agent/config/config.go]: Done
